**Nome: Ozeas dos Santos Nascimento**


A base de dados automoveis.xlsx necessita de tratamento. 

Carregarei a base de dados Automoveis.xlsx, e realizarei a preparação dos dados para que eles sejam consistentes. Neste, é possivel encontrar:
Valores omissos que poderão ser registros ou atributos.
Valores omissos que poderão ser estimados.
Valores inconsistentes.
Problemas de formatação de dados.


In [ ]:
import pandas as pd
import matplotlib as plt
import numpy as np

### Leitura dos Dados - Brutos

In [ ]:
df_auto = pd.read_excel("automoveis.xls" )
df_auto

Fabricante Combustível  Portas Estilo Chassis     Tração Comprimento  \
0    alfa-romeo         gas     2.0    conversível   traseira      168,80   
1    alfa-romeo         gas     2.0    conversível   traseira      168,80   
2    alfa-romeo         gas     2.0          hatch   traseira      171,20   
3          audi         gas     4.0          sedan  dianteira      176,60   
4          audi         gas     4.0          sedan        4x4      176,60   
..          ...         ...     ...            ...        ...         ...   
200       volvo         gas     4.0          sedan   traseira      188,80   
201       volvo      diesel     4.0          sedan   traseira      188,80   
202       volvo         gas     4.0          sedan   traseira      188,80   
203         NaN         gas     4.0          sedan   traseira      176,80   
204         NaN         gas     2.0          hatch   traseira      169,00   

    Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
0     64,10  48,80          dohc                   4           130,00   
1     64,10  48,80          dohc                   4           130,00   
2     65,50  52,40          ohcv                   6           152,00   
3     66,20  54,30           ohc                   4           109,00   
4     66,40  54,30           ohc                   5           136,00   
..      ...    ...           ...                 ...              ...   
200   68,90  55,50          ohcv                   6           173,00   
201   68,90  55,50           ohc                   6           145,00   
202   68,90  55,50           ohc                   4           141,00   
203   64,80  54,30           ohc                   6           164,00   
204   65,70  49,60         rotor                   2            70,00   

    Tipo de injeção Potência (HP) Pico RPM     Preço  
0              mpfi        111,00  5000,00  13495,00  
1              mpfi        111,00  5000,00  16500,00  
2              mpfi        154,00  5000,00  16500,00  
3              mpfi        102,00  5500,00  13950,00  
4              mpfi        115,00  5500,00  17450,00  
..              ...           ...      ...       ...  
200            mpfi        134,00  5500,00  21485,00  
201             idi        106,00  4800,00  22470,00  
202            mpfi        114,00  5400,00  22625,00  
203            mpfi        121,00  4250,00  21105,00  
204            4bbl        101,00  6000,00  10945,00  

[205 rows x 15 columns]

## Analise Tipo de Dados - usarei a função info() para analisar como os dados estão disponíveis
### Temos 205 entradas de dados, e o total de 15 colunas, desta, Apenas a coluna "Portas" está com o tipo float, 

In [ ]:
df_auto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 15 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Fabricante           203 non-null    object 
 1   Combustível          200 non-null    object 
 2   Portas               203 non-null    float64
 3   Estilo Chassis       205 non-null    object 
 4   Tração               205 non-null    object 
 5   Comprimento          205 non-null    object 
 6   Largura              205 non-null    object 
 7   Altura               205 non-null    object 
 8   Tipo de motor        205 non-null    object 
 9   Número de cilindros  205 non-null    object 
 10  Tamanho do motor     205 non-null    object 
 11  Tipo de injeção      205 non-null    object 
 12  Potência (HP)        205 non-null    object 
 13  Pico RPM             205 non-null    object 
 14  Preço                205 non-null    object 
dtypes: float64(1), object(14)
memory usage: 

## Identificação Dados NA
### Identificarei a quantidade de dados faltantes em nosso dataset

In [ ]:
df_auto.isna().sum()

Fabricante             2
Combustível            5
Portas                 2
Estilo Chassis         0
Tração                 0
Comprimento            0
Largura                0
Altura                 0
Tipo de motor          0
Número de cilindros    0
Tamanho do motor       0
Tipo de injeção        0
Potência (HP)          0
Pico RPM               0
Preço                  0
dtype: int64

## Identificação Dados Ausentes Fabricantes
### Como temos poucos dados (205), optei por analisar todo o dataset, afim de encontrar algum padrão para poder inferir sobre os dados faltantes

In [ ]:
df_auto.tail()

Fabricante Combustível  Portas Estilo Chassis    Tração Comprimento  \
200      volvo         gas     4.0          sedan  traseira      188,80   
201      volvo      diesel     4.0          sedan  traseira      188,80   
202      volvo         gas     4.0          sedan  traseira      188,80   
203        NaN         gas     4.0          sedan  traseira      176,80   
204        NaN         gas     2.0          hatch  traseira      169,00   

    Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
200   68,90  55,50          ohcv                   6           173,00   
201   68,90  55,50           ohc                   6           145,00   
202   68,90  55,50           ohc                   4           141,00   
203   64,80  54,30           ohc                   6           164,00   
204   65,70  49,60         rotor                   2            70,00   

    Tipo de injeção Potência (HP) Pico RPM     Preço  
200            mpfi        134,00  5500,00  21485,00  
201             idi        106,00  4800,00  22470,00  
202            mpfi        114,00  5400,00  22625,00  
203            mpfi        121,00  4250,00  21105,00  
204            4bbl        101,00  6000,00  10945,00

In [ ]:
mot = df_auto.loc[df_auto["Número de cilindros"]== 6 ]
mot= mot.loc[mot["Tipo de motor"]=="ohc"]
mot = mot.loc[mot["Combustível"]=="gas"]
mot



Fabricante Combustível  Portas Estilo Chassis    Tração Comprimento  \
12         bmw         gas     2.0          sedan  traseira      176,80   
13         bmw         gas     4.0          sedan  traseira      189,00   
14         bmw         gas     4.0          sedan  traseira      189,00   
15         bmw         gas     2.0          sedan  traseira      193,80   
16         bmw         gas     4.0          sedan  traseira      197,00   
203        NaN         gas     4.0          sedan  traseira      176,80   

    Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
12    64,80  54,30           ohc                   6           164,00   
13    66,90  55,70           ohc                   6           164,00   
14    66,90  55,70           ohc                   6           209,00   
15    67,90  53,70           ohc                   6           209,00   
16    70,90  56,30           ohc                   6           209,00   
203   64,80  54,30           ohc                   6           164,00   

    Tipo de injeção Potência (HP) Pico RPM     Preço  
12             mpfi        121,00  4250,00  20970,00  
13             mpfi        121,00  4250,00  24565,00  
14             mpfi        182,00  5400,00  30760,00  
15             mpfi        182,00  5400,00  41315,00  
16             mpfi        182,00  5400,00  36880,00  
203            mpfi        121,00  4250,00  21105,00

In [ ]:
mot = df_auto.loc[df_auto["Número de cilindros"]== 2 ]
mot= mot.loc[mot["Tipo de motor"]=="rotor"]
mot

Fabricante Combustível  Portas Estilo Chassis    Tração Comprimento  \
54       mazda         gas     2.0          hatch  traseira      169,00   
55       mazda         gas     2.0          hatch  traseira      169,00   
56       mazda         gas     2.0          hatch  traseira      169,00   
204        NaN         gas     2.0          hatch  traseira      169,00   

    Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
54    65,70  49,60         rotor                   2            70,00   
55    65,70  49,60         rotor                   2            70,00   
56    65,70  49,60         rotor                   2            80,00   
204   65,70  49,60         rotor                   2            70,00   

    Tipo de injeção Potência (HP) Pico RPM     Preço  
54             4bbl        101,00  6000,00  11845,00  
55             4bbl        101,00  6000,00  13645,00  
56             mpfi        135,00  6000,00  15645,00  
204            4bbl        101,00  6000,00  10945,00

## Substituindo dados ausentes
### Como previsto, foi possivel inferir sobre os dados faltantes. farei agora os devidos ajustes

In [ ]:
df_bmw = df_auto.loc[df_auto["Fabricante"].index == 203].replace(np.nan,"bmw")
df_bmw

Fabricante Combustível  Portas Estilo Chassis    Tração Comprimento  \
203        bmw         gas     4.0          sedan  traseira      176,80   

    Largura Altura Tipo de motor  Número de cilindros Tamanho do motor  \
203   64,80  54,30           ohc                    6           164,00   

    Tipo de injeção Potência (HP) Pico RPM     Preço  
203            mpfi        121,00  4250,00  21105,00

In [ ]:
df_maz = df_auto.loc[df_auto["Fabricante"].index == 204].replace(np.nan,"mazda")
df_maz

Fabricante Combustível  Portas Estilo Chassis    Tração Comprimento  \
204      mazda         gas     2.0          hatch  traseira      169,00   

    Largura Altura Tipo de motor  Número de cilindros Tamanho do motor  \
204   65,70  49,60         rotor                    2            70,00   

    Tipo de injeção Potência (HP) Pico RPM     Preço  
204            4bbl        101,00  6000,00  10945,00

## JOIN  
### Hora de juntar os dataframes com as informações corretas

In [ ]:
#Deletando linhas ausentes pelo indice
df_auto = df_auto.drop(203)
df_auto = df_auto.drop(204)
#Join DFs
dfjoin = [df_auto, df_bmw, df_maz]
df_auto = pd.concat(dfjoin)

In [ ]:
df_auto.tail()

Fabricante Combustível  Portas Estilo Chassis    Tração Comprimento  \
200      volvo         gas     4.0          sedan  traseira      188,80   
201      volvo      diesel     4.0          sedan  traseira      188,80   
202      volvo         gas     4.0          sedan  traseira      188,80   
203        bmw         gas     4.0          sedan  traseira      176,80   
204      mazda         gas     2.0          hatch  traseira      169,00   

    Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
200   68,90  55,50          ohcv                   6           173,00   
201   68,90  55,50           ohc                   6           145,00   
202   68,90  55,50           ohc                   4           141,00   
203   64,80  54,30           ohc                   6           164,00   
204   65,70  49,60         rotor                   2            70,00   

    Tipo de injeção Potência (HP) Pico RPM     Preço  
200            mpfi        134,00  5500,00  21485,00  
201             idi        106,00  4800,00  22470,00  
202            mpfi        114,00  5400,00  22625,00  
203            mpfi        121,00  4250,00  21105,00  
204            4bbl        101,00  6000,00  10945,00

## Tratamento dados ausentes em Combustiveis / / Comparações de Dados
Realizarei a mesma operação

In [ ]:
df_auto.loc[df_auto["Combustível"].isnull()]


Fabricante Combustível  Portas Estilo Chassis     Tração Comprimento  \
117   plymouth         NaN     2.0          hatch  dianteira      157,30   
118   plymouth         NaN     4.0          hatch  dianteira      157,30   
119   plymouth         NaN     4.0          sedan  dianteira      167,30   
125    porsche         NaN     2.0        hardtop   traseira      168,90   
139     subaru         NaN     4.0          sedan  dianteira      172,00   

    Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
117   63,80  50,80           ohc                   4            98,00   
118   63,80  50,60           ohc                   4            90,00   
119   63,80  50,80           ohc                   4            90,00   
125   65,00  51,60          ohcf                   6           194,00   
139   65,40  52,50          ohcf                   4           108,00   

    Tipo de injeção Potência (HP) Pico RPM     Preço  
117            spdi        102,00  5500,00   7957,00  
118            2bbl         68,00  5500,00   6229,00  
119            2bbl         68,00  5500,00   6692,00  
125            mpfi        207,00  5900,00  34028,00  
139            2bbl         82,00  4800,00   7126,00

In [ ]:
df_auto.loc[df_auto["Fabricante"] == "plymouth"]

Fabricante Combustível  Portas Estilo Chassis     Tração Comprimento  \
116   plymouth         gas     2.0          hatch  dianteira      157,30   
117   plymouth         NaN     2.0          hatch  dianteira      157,30   
118   plymouth         NaN     4.0          hatch  dianteira      157,30   
119   plymouth         NaN     4.0          sedan  dianteira      167,30   
120   plymouth         gas     4.0          sedan  dianteira      167,30   
121   plymouth         gas     4.0            suv  dianteira      174,60   
122   plymouth         gas     2.0          hatch   traseira      173,20   

    Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
116   63,80  50,80           ohc                   4            90,00   
117   63,80  50,80           ohc                   4            98,00   
118   63,80  50,60           ohc                   4            90,00   
119   63,80  50,80           ohc                   4            90,00   
120   63,80  50,80           ohc                   4            98,00   
121   64,60  59,80           ohc                   4           122,00   
122   66,30  50,20           ohc                   4           156,00   

    Tipo de injeção Potência (HP) Pico RPM     Preço  
116            2bbl         68,00  5500,00   5572,00  
117            spdi        102,00  5500,00   7957,00  
118            2bbl         68,00  5500,00   6229,00  
119            2bbl         68,00  5500,00   6692,00  
120            2bbl         68,00  5500,00   7609,00  
121            2bbl         88,00  5000,00   8921,00  
122            spdi        145,00  5000,00  12764,00

In [ ]:
df_auto.loc[df_auto["Fabricante"] == "porsche"]

Fabricante Combustível  Portas Estilo Chassis    Tração Comprimento  \
123    porsche         gas     2.0          hatch  traseira      168,90   
124    porsche         gas     2.0        hardtop  traseira      168,90   
125    porsche         NaN     2.0        hardtop  traseira      168,90   
126    porsche         gas     2.0    conversível  traseira      168,90   
127    porsche         gas     2.0          hatch  traseira      175,70   

    Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
123   68,30  50,20           ohc                   4           151,00   
124   65,00  51,60          ohcf                   6           194,00   
125   65,00  51,60          ohcf                   6           194,00   
126   65,00  51,60          ohcf                   6           194,00   
127   72,30  50,50         dohcv                oito           203,00   

    Tipo de injeção Potência (HP) Pico RPM     Preço  
123            mpfi        143,00  5500,00  22018,00  
124            mpfi        207,00  5900,00  32528,00  
125            mpfi        207,00  5900,00  34028,00  
126            mpfi        207,00  5900,00  37028,00  
127            mpfi        288,00  5750,00  13207,13

In [ ]:
df_auto.loc[df_auto["Fabricante"] == "subaru"]

Fabricante Combustível  Portas Estilo Chassis     Tração Comprimento  \
136     subaru         gas     2.0          hatch  dianteira      156,90   
137     subaru         gas     2.0          hatch  dianteira      157,90   
138     subaru         gas     2.0          hatch        4x4      157,30   
139     subaru         NaN     4.0          sedan  dianteira      172,00   
140     subaru         gas     4.0          sedan  dianteira      172,00   
141     subaru         gas     4.0          sedan  dianteira      172,00   
142     subaru         gas     4.0          sedan        4x4      172,00   
143     subaru         gas     4.0          sedan        4x4      172,00   
144     subaru         gas     4.0            suv  dianteira      173,50   
145     subaru         gas     4.0            suv  dianteira      173,50   
146     subaru         gas     4.0            suv        4x4      173,60   
147     subaru         gas     4.0            suv        4x4      173,60   

    Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
136   63,40  53,70          ohcf                   4            97,00   
137   63,60  53,70          ohcf                   4           108,00   
138   63,80  55,70          ohcf                   4           108,00   
139   65,40  52,50          ohcf                   4           108,00   
140   65,40  52,50          ohcf                   4           108,00   
141   65,40  52,50          ohcf                   4           108,00   
142   65,40  54,30          ohcf                   4           108,00   
143   65,40  54,30          ohcf                   4           108,00   
144   65,40  53,00          ohcf                   4           108,00   
145   65,40  53,00          ohcf                   4           108,00   
146   65,40  54,90          ohcf                   4           108,00   
147   65,40  54,90          ohcf                   4           108,00   

    Tipo de injeção Potência (HP) Pico RPM     Preço  
136            2bbl         69,00  4900,00   5118,00  
137            2bbl         73,00  4400,00   7053,00  
138            2bbl         73,00  4400,00   7603,00  
139            2bbl         82,00  4800,00   7126,00  
140            2bbl         82,00  4400,00   7775,00  
141            mpfi         94,00  5200,00   9960,00  
142            2bbl         82,00  4800,00   9233,00  
143            mpfi        111,00  4800,00  11259,00  
144            2bbl         82,00  4800,00   7463,00  
145            mpfi         94,00  5200,00  10198,00  
146            2bbl         82,00  4800,00   8013,00  
147            mpfi        111,00  4800,00  11694,00

In [ ]:
df_auto["Combustível"] = df_auto["Combustível"].fillna("gas")

In [ ]:
df_auto

Fabricante Combustível  Portas Estilo Chassis     Tração Comprimento  \
0    alfa-romeo         gas     2.0    conversível   traseira      168,80   
1    alfa-romeo         gas     2.0    conversível   traseira      168,80   
2    alfa-romeo         gas     2.0          hatch   traseira      171,20   
3          audi         gas     4.0          sedan  dianteira      176,60   
4          audi         gas     4.0          sedan        4x4      176,60   
..          ...         ...     ...            ...        ...         ...   
200       volvo         gas     4.0          sedan   traseira      188,80   
201       volvo      diesel     4.0          sedan   traseira      188,80   
202       volvo         gas     4.0          sedan   traseira      188,80   
203         bmw         gas     4.0          sedan   traseira      176,80   
204       mazda         gas     2.0          hatch   traseira      169,00   

    Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
0     64,10  48,80          dohc                   4           130,00   
1     64,10  48,80          dohc                   4           130,00   
2     65,50  52,40          ohcv                   6           152,00   
3     66,20  54,30           ohc                   4           109,00   
4     66,40  54,30           ohc                   5           136,00   
..      ...    ...           ...                 ...              ...   
200   68,90  55,50          ohcv                   6           173,00   
201   68,90  55,50           ohc                   6           145,00   
202   68,90  55,50           ohc                   4           141,00   
203   64,80  54,30           ohc                   6           164,00   
204   65,70  49,60         rotor                   2            70,00   

    Tipo de injeção Potência (HP) Pico RPM     Preço  
0              mpfi        111,00  5000,00  13495,00  
1              mpfi        111,00  5000,00  16500,00  
2              mpfi        154,00  5000,00  16500,00  
3              mpfi        102,00  5500,00  13950,00  
4              mpfi        115,00  5500,00  17450,00  
..              ...           ...      ...       ...  
200            mpfi        134,00  5500,00  21485,00  
201             idi        106,00  4800,00  22470,00  
202            mpfi        114,00  5400,00  22625,00  
203            mpfi        121,00  4250,00  21105,00  
204            4bbl        101,00  6000,00  10945,00  

[205 rows x 15 columns]

### Verificação dados ausentes - Portas
Durante a verificação, foi identificado que todos os sedans possuem 4.0 portas como padrão, não sendo necessario a exclusão das linhas ausentes.

In [ ]:
df_auto.loc[df_auto["Portas"].isnull()]

Fabricante Combustível  Portas Estilo Chassis     Tração Comprimento  \
26      dodge         gas     NaN          sedan  dianteira      157,30   
61      mazda      diesel     NaN          sedan  dianteira      177,80   

   Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
26   63,80  50,60           ohc                   4            98,00   
61   66,50  55,50           ohc                   4           122,00   

   Tipo de injeção Potência (HP) Pico RPM     Preço  
26            mpfi        102,00  5500,00   8558,00  
61             idi         64,00  4650,00  10795,00

In [ ]:
df_auto.loc[(df_auto["Fabricante"]== "dodge")]

Fabricante Combustível  Portas Estilo Chassis     Tração Comprimento  \
20      dodge         gas     2.0          hatch  dianteira      157,30   
21      dodge         gas     2.0          hatch  dianteira      157,30   
22      dodge         gas     2.0          hatch  dianteira      157,30   
23      dodge         gas     4.0          hatch  dianteira      157,30   
24      dodge         gas     4.0          sedan  dianteira      157,30   
25      dodge         gas     4.0          sedan  dianteira      157,30   
26      dodge         gas     NaN          sedan  dianteira      157,30   
27      dodge         gas     4.0            suv  dianteira      174,60   
28      dodge         gas     2.0          hatch  dianteira      173,20   

   Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
20   63,80  50,80           ohc                   4            90,00   
21   63,80  50,80           ohc                   4            90,00   
22   63,80  50,80           ohc                   4            98,00   
23   63,80  50,60           ohc                   4            90,00   
24   63,80  50,60           ohc                   4            90,00   
25   63,80  50,60           ohc                   4            90,00   
26   63,80  50,60           ohc                   4            98,00   
27   64,60  59,80           ohc                   4           122,00   
28   66,30  50,20           ohc                   4           156,00   

   Tipo de injeção Potência (HP) Pico RPM     Preço  
20            2bbl         68,00  5500,00   5572,00  
21            2bbl         68,00  5500,00   6377,00  
22            mpfi        102,00  5500,00   7957,00  
23            2bbl         68,00  5500,00   6229,00  
24            2bbl         68,00  5500,00   6692,00  
25            2bbl         68,00  5500,00   7609,00  
26            mpfi        102,00  5500,00   8558,00  
27            2bbl         88,00  5000,00   8921,00  
28             mfi        145,00  5000,00  12964,00

In [ ]:
df_auto.loc[(df_auto["Fabricante"]== "mazda")]

Fabricante Combustível  Portas Estilo Chassis     Tração Comprimento  \
49       mazda         gas     2.0          hatch  dianteira      159,10   
50       mazda         gas     2.0          hatch  dianteira      159,10   
51       mazda         gas     2.0          hatch  dianteira      159,10   
52       mazda         gas     4.0          sedan  dianteira      166,80   
53       mazda         gas     4.0          sedan  dianteira      166,80   
54       mazda         gas     2.0          hatch   traseira      169,00   
55       mazda         gas     2.0          hatch   traseira      169,00   
56       mazda         gas     2.0          hatch   traseira      169,00   
57       mazda         gas     2.0          hatch  dianteira      177,80   
58       mazda         gas     4.0          sedan  dianteira      177,80   
59       mazda         gas     2.0          hatch  dianteira      177,80   
60       mazda         gas     4.0          sedan  dianteira      177,80   
61       mazda      diesel     NaN          sedan  dianteira      177,80   
62       mazda         gas     4.0          hatch  dianteira      177,80   
63       mazda         gas     4.0          sedan   traseira      175,00   
64       mazda      diesel     4.0          sedan   traseira      175,00   
204      mazda         gas     2.0          hatch   traseira      169,00   

    Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
49    64,20  54,10           ohc                   4            91,00   
50    64,20  54,10           ohc                   4            91,00   
51    64,20  54,10           ohc                   4            91,00   
52    64,20  54,10           ohc                   4            91,00   
53    64,20  54,10           ohc                   4            91,00   
54    65,70  49,60         rotor                   2            70,00   
55    65,70  49,60         rotor                   2            70,00   
56    65,70  49,60         rotor                   2            80,00   
57    66,50  53,70           ohc                   4           122,00   
58    66,50  55,50           ohc                   4           122,00   
59    66,50  53,70           ohc                   4           122,00   
60    66,50  55,50           ohc                   4           122,00   
61    66,50  55,50           ohc                   4           122,00   
62    66,50  55,50           ohc                   4           122,00   
63    66,10  54,40           ohc                   4           140,00   
64    66,10  54,40           ohc                   4           134,00   
204   65,70  49,60         rotor                   2            70,00   

    Tipo de injeção Potência (HP) Pico RPM     Preço  
49             2bbl         68,00  5000,00   5195,00  
50             2bbl         68,00  5000,00   6095,00  
51             2bbl         68,00  5000,00   6795,00  
52             2bbl         68,00  5000,00   6695,00  
53             2bbl         68,00  5000,00   7395,00  
54             4bbl        101,00  6000,00  11845,00  
55             4bbl        101,00  6000,00  13645,00  
56             mpfi        135,00  6000,00  15645,00  
57             2bbl         84,00  4800,00   8845,00  
58             2bbl         84,00  4800,00   8495,00  
59             2bbl         84,00  4800,00  10595,00  
60             2bbl         84,00  4800,00  10245,00  
61              idi         64,00  4650,00  10795,00  
62             2bbl         84,00  4800,00  11245,00  
63             mpfi        120,00  5000,00  18280,00  
64              idi         72,00  4200,00  18344,00  
204            4bbl        101,00  6000,00  10945,00

In [ ]:
df_auto["Portas"] = df_auto["Portas"].fillna(4.0)

In [ ]:
df_auto.isnull().sum()

Fabricante             0
Combustível            0
Portas                 0
Estilo Chassis         0
Tração                 0
Comprimento            0
Largura                0
Altura                 0
Tipo de motor          0
Número de cilindros    0
Tamanho do motor       0
Tipo de injeção        0
Potência (HP)          0
Pico RPM               0
Preço                  0
dtype: int64

### Conforme podemos analisar na celula anterior, não possuimos mais dados faltantes em nosso DF. Seguirei agora para o tratamento dos tipos de dados
## Tratamento Tipo de Dados

In [ ]:
df_auto.head(2)

Fabricante Combustível  Portas Estilo Chassis    Tração Comprimento  \
0  alfa-romeo         gas     2.0    conversível  traseira      168,80   
1  alfa-romeo         gas     2.0    conversível  traseira      168,80   

  Largura Altura Tipo de motor Número de cilindros Tamanho do motor  \
0   64,10  48,80          dohc                   4           130,00   
1   64,10  48,80          dohc                   4           130,00   

  Tipo de injeção Potência (HP) Pico RPM     Preço  
0            mpfi        111,00  5000,00  13495,00  
1            mpfi        111,00  5000,00  16500,00

In [ ]:
df_auto.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 205 entries, 0 to 204
Data columns (total 15 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Fabricante           205 non-null    object 
 1   Combustível          205 non-null    object 
 2   Portas               205 non-null    float64
 3   Estilo Chassis       205 non-null    object 
 4   Tração               205 non-null    object 
 5   Comprimento          205 non-null    object 
 6   Largura              205 non-null    object 
 7   Altura               205 non-null    object 
 8   Tipo de motor        205 non-null    object 
 9   Número de cilindros  205 non-null    object 
 10  Tamanho do motor     205 non-null    object 
 11  Tipo de injeção      205 non-null    object 
 12  Potência (HP)        205 non-null    object 
 13  Pico RPM             205 non-null    object 
 14  Preço                205 non-null    object 
dtypes: float64(1), object(14)
memory usage: 

In [ ]:
df_auto["Número de cilindros"].unique()

array([4, 6, 5, 3, 'doze', 2, 'oito'], dtype=object)

In [ ]:
df_auto["Preço"].unique()

array(['13495,00', '16500,00', '13950,00', '17450,00', '15250,00',
       '17710,00', '18920,00', '23875,00', '13207,13', '16430,00',
       '16925,00', '20970,00', '24565,00', '30760,00', '41315,00',
       '36880,00', '5151,00', '6295,00', '6575,00', '5572,00', '6377,00',
       '7957,00', '6229,00', '6692,00', '7609,00', '8558,00', '8921,00',
       '12964,00', '6479,00', '6855,00', '5399,00', '6529,00', '7129,00',
       '7295,00', '7895,00', '9095,00', '8845,00', '10295,00', '12945,00',
       '10345,00', '6785,00', '11048,00', '32250,00', '35550,00',
       '36000,00', '5195,00', '6095,00', '6795,00', '6695,00', '7395,00',
       '11845,00', '13645,00', '15645,00', '8495,00', '10595,00',
       '10245,00', '10795,00', '11245,00', '18280,00', '18344,00',
       '25552,00', '28248,00', '28176,00', '31600,00', '34184,00',
       '35056,00', '40960,00', '45400,00', '16503,00', '5389,00',
       '6189,00', '6669,00', '7689,00', '9959,00', '8499,00', '12629,00',
       '14869,00', '144

In [ ]:
df_auto["Pico RPM"].unique()

array(['5000,00', '5500,00', 580000, '4250,00', '5400,00', '5100,00',
       '4800,00', '6000,00', '5800,00', '4750,00', '4650,00', '4200,00',
       '4350,00', '4500,00', '5200,00', '4150,00', '5600,00', '5900,00',
       '5750,00', '5125,37', '5250,00', '4900,00', '4400,00', '6600,00',
       '5300,00'], dtype=object)

In [ ]:
df_auto["Potência (HP)"].unique()

array(['111,00', '154,00', '102,00', '115,00', '110,00', '140,00',
       '160,00', '101,00', '121,00', '182,00', '48,00', '70,00', '68,00',
       '88,00', '145,00', '58,00', '76,00', '60,00', '86,00', '100,00',
       '78,00', '90,00', '176,00', '262,00', '135,00', '84,00', '64,00',
       '120,00', '72,00', '123,00', '155,00', '184,00', '175,00',
       '116,00', '69,00', '55,00', '97,00', '152,00', '200,00', '95,00',
       '142,00', '143,00', '207,00', '288,00', '104,26', '73,00', '82,00',
       '94,00', '62,00', '56,00', '112,00', '92,00', '161,00', '156,00',
       '52,00', '85,00', '114,00', '162,00', '134,00', '106,00'],
      dtype=object)

In [ ]:
df_auto["Comprimento"].unique()

array(['168,80', '171,20', '176,60', '177,30', '192,70', '178,20',
       '176,80', '189,00', '193,80', '197,00', '141,10', '155,90',
       '158,80', '157,30', '174,60', '173,20', '144,60', '150,00',
       '163,40', '157,10', '167,50', '175,40', '169,10', '170,70',
       '172,60', '199,60', '191,70', '159,10', '166,80', '169,00',
       '177,80', '175,00', '190,90', '187,50', '202,60', '180,30',
       '208,10', '199,20', '178,40', '173,00', '172,40', '165,30',
       '170,20', '165,60', '162,40', '173,40', '181,70', '184,60',
       '178,50', '186,70', '198,90', '167,30', '168,90', '175,70',
       '181,50', '186,60', '156,90', '157,90', '172,00', '173,50',
       '173,60', '158,70', '169,70', '166,30', '168,70', '176,20',
       '175,60', '183,50', '187,80', '171,70', '159,30', '165,70',
       '180,20', '183,10', '188,80'], dtype=object)

In [ ]:
df_auto["Tamanho do motor"].unique()

array(['130,00', '152,00', '109,00', '136,00', '131,00', '108,00',
       '164,00', '209,00', '61,00', '90,00', '98,00', '122,00', '156,00',
       '92,00', '79,00', '110,00', '111,00', '119,00', '258,00', '326,00',
       '91,00', '70,00', '80,00', '140,00', '134,00', '183,00', '234,00',
       '308,00', '304,00', '97,00', '103,00', '120,00', '181,00',
       '151,00', '194,00', '203,00', '132,00', '121,00', '146,00',
       '171,00', '161,00', '141,00', '173,00', '145,00'], dtype=object)

In [ ]:
df_auto["Preço"] = df_auto["Preço"].replace('\.','', regex=True).replace(',','.', regex=True).astype(float)
df_auto["Tamanho do motor"] = df_auto["Tamanho do motor"].replace('\.','', regex=True).replace(',','.', regex=True).astype(float)
df_auto["Comprimento"] = df_auto["Comprimento"].replace('\.','', regex=True).replace(',','.', regex=True).astype(float)
df_auto["Largura"] = df_auto["Largura"].replace('\.','', regex=True).replace(',','.', regex=True).astype(float)
df_auto["Altura"] = df_auto["Altura"].replace('\.','', regex=True).replace(',','.', regex=True).astype(float)
df_auto["Potência (HP)"] = df_auto["Potência (HP)"].replace('\.','', regex=True).replace(',','.', regex=True).astype(float)
df_auto["Portas"] = df_auto["Portas"].replace(",",".").astype(float)
df_auto["Número de cilindros"] = df_auto["Número de cilindros"].replace("doze",12).replace("oito",8).astype(float)
df_auto["Pico RPM"] = df_auto["Pico RPM"].replace('\.','', regex=True).replace(',','.', regex=True).astype(float)

In [ ]:
df_auto.head(2)

Fabricante Combustível  Portas Estilo Chassis    Tração  Comprimento  \
0  alfa-romeo         gas     2.0    conversível  traseira        168.8   
1  alfa-romeo         gas     2.0    conversível  traseira        168.8   

   Largura  Altura Tipo de motor  Número de cilindros  Tamanho do motor  \
0     64.1    48.8          dohc                  4.0             130.0   
1     64.1    48.8          dohc                  4.0             130.0   

  Tipo de injeção  Potência (HP)  Pico RPM    Preço  
0            mpfi          111.0    5000.0  13495.0  
1            mpfi          111.0    5000.0  16500.0

In [ ]:
df_auto.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 205 entries, 0 to 204
Data columns (total 15 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Fabricante           205 non-null    object 
 1   Combustível          205 non-null    object 
 2   Portas               205 non-null    float64
 3   Estilo Chassis       205 non-null    object 
 4   Tração               205 non-null    object 
 5   Comprimento          205 non-null    float64
 6   Largura              205 non-null    float64
 7   Altura               205 non-null    float64
 8   Tipo de motor        205 non-null    object 
 9   Número de cilindros  205 non-null    float64
 10  Tamanho do motor     205 non-null    float64
 11  Tipo de injeção      205 non-null    object 
 12  Potência (HP)        205 non-null    float64
 13  Pico RPM             205 non-null    float64
 14  Preço                205 non-null    float64
dtypes: float64(9), object(6)
memory usage: 2

### Todo o dataset foi tratado, com os dados corretamente ajustados e sem dados faltantes.